packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -64,85 +64,54 @@ export const ultraFeedGraphQLTypeDefs = gql`
   }
 `
 
+// items now carry `type`
+type SampledItem =
+  | { type: "feedCommentThread"; feedCommentThread: FeedCommentsThread }
+  | { type: "feedPostWithContents"; feedPost: FeedFullPost }
+  | { type: "feedPost"; feedPostStub: FeedPostStub }          // ← NEW stub
+  | { type: "feedSpotlight"; feedSpotlight: FeedSpotlight };
+
 interface WeightedSource {
   weight: number;
-  items: FeedItem[];
-  renderAsType: FeedItemRenderType;
+  items: SampledItem[];
 }
 
-type SampledItem = { renderAsType: "feedCommentThread", feedCommentThread: FeedCommentsThread }
-                 | { renderAsType: "feedPost", feedPost: FeedFullPost }
-                 | { renderAsType: "feedSpotlight", feedSpotlight: FeedSpotlight };
-
 const weightedSample = (
-  inputs: Record<FeedItemSourceType, WeightedSource>,
+  inputs: Partial<Record<FeedItemSourceType, WeightedSource>>,
   totalItems: number
 ): SampledItem[] => {
-  // Create deep copies of the input arrays to avoid modifying the originals
   const sourcesWithCopiedItems = cloneDeep(inputs);
-
   const finalFeed: SampledItem[] = [];
-  let totalWeight = Object.values(sourcesWithCopiedItems).reduce(
-    (sum, src) => sum + (src.items.length > 0 ? src.weight : 0),
-    0
-  );
+
+  let totalWeight = Object.values(sourcesWithCopiedItems)
+    .reduce((sum, s) => sum + ((s && s.items.length) ? s.weight : 0), 0);
```

Comment:
if it really does need to be partial, then `sum + (s?.items.length ? s.weight : 0)`